### ISSUE TITLE
Restrict data transfer options for the secure data storages

### ISSUE THREAD

**Issue by sidoruka** [epam/cloud-pipeline#1036]
**Background**
At the moment, Cloud Pipeline allows performing upload/download operations for any authorized data storage.
But certain storages may contain sensitive data, which shall not be copied anywhere outside that storage.
**_Shall be considered for the object storage only_**

**Approach**
* We shall list all the available "data-leak" routes (e.g. download from the GUI, copy to another bucket within an instance via mounts, copy via the CLI, upload to some external resource, etc.) and the ways they can be restricted
* Add a boolean configuration option to the object storages (e.g. `Sensitive storage`)
* For all the storages, marked as `Sensitive` - enforce the "data-leak" prevention rules, which are going to be listed in (1)

**Comment by rodichenko** [comment]
As for the "data-leak" routes (GUI):

## Cloud Pipeline
Data storage objects upload:
- `storage/<ID>` route: **Upload** button

Data storage objects download & content-fetch:
- `storage/<ID>` route:
  - **Download** button;
  - **Generate URL** button
  -  Object's metadata window with **Download file** button (inline file preview)
  - **Preview** file contents window with **Download file** button
- Global search:
  - **Download file to view full contents** button for Object Storage Item preview

## FS Browser
 - **Download** object button
 - **Upload** area

**Comment by sidoruka** [comment]
"Data leak" routes and restrictriction options:
* Internet (NAT/ExternalIP/http_proxy)
  * Let's allow to configure a separate "sensitive" subnet(s) 
  * It shall be used to create new nodes for runs, which have at least one sensitive bucket attached (i.e. if a `CP_CAP_LIMIT_MOUNTS` contains at least one storage id, which is "sensitive")
* GUI download
  * Hide any `Download` or `Generate URL` button in the `sensitive` bucket browser
  * API shall not process any request for `Generate URL` requests for the `sensitive` buckets
* `pipe cp`
  * From **outside** the "sensitive" run (which has at least one sensitive bucket attached)
    * Do not allow any `cp/mv/rm` operations, where the source or destination is a `sensitive` bucket
    * `ls` operation shall be still available
  * From **inside** the "sensitive" run
    * `cp` operation shall be available to download the data to the run's filesystem
    * Other operations shall work as previous item (**outside** the "sensitive" run)
* `pipe storage mount -b`
  * From **outside** the "sensitive" run (which has at least one sensitive bucket attached)
    * Do not allow to mount `sensitive` buckets (inform the user and exit)
  * From **inside** the "sensitive" run
    * Allow to mount all the buckets (including `sensitive`)  in a `readonly` mode
* Copy to another mounted bucket within a running job
  * All the storages for the job, which has `sensitive` buckets attached, shall be mounted in a read-only mode
  * If `CP_CAP_LIMIT_MOUNTS` is not defined - all "non-sensitive" buckets shall be mounted (at the moment, this means that ALL the buckets are mounted)
* Download via the interactive endpoint GUI
  * Restrict data transfer capabilities of the major tools (via some proxy config):
    * `rstudio`
    * `jupyter (notebook/lab)`
    * `nomachine`
  * Specific requirements are listed in #1175
* FSBrowser
  * Do not install/configure FSBrowser for the jobs, which have `sensitive` buckets attached
  * Do not show FSBrowser button in the GUI for such jobs
* `pipe scp` (to be implemented via #640)
  * TBD, once #640 is done

Notes on how to determine if we are running pipe `inside` or `outside` the job:
* The sensitive runs are launched in the `sensitive` subnets
* We can get the IP ranges, from those subnets
* Only the requests, received from these subnets, shall be served by the API

Notes on the `sensitive` networks:
* We shall be able to setup a separate subnet in the kube's overlay network
* "Sensitive" runs shall be launched in this network if a certain label is set in pod's spec
* This network shall allow to limit the ingress/egress connection

**Comment by mzueva** [comment]
@Wedds Let's proceed with the first part of server-side implementation:
1. Add boolean `sensitive` attribute to `AbstractDatastorage` entity and persist to the database
2. Restrict all server methods to get file content from `senstitive` buckets

**Comment by rodichenko** [comment]
@sidoruka GUI part implemented via 9c2f9e1fb273ce29b518d7cef55f2410c0cbb5f6

**Comment by rodichenko** [comment]
@sidoruka GUI part backported to `release/0.15` via d17ebe1845b22fc7ab84b3c4b24b799290ee5f28

**Comment by mzueva** [comment]
@sidoruka handling of `sensitive` runs is described in a new issue #1124

**Comment by NShaforostov** [comment]
I've tried to verify the feature.
In general, it works well.
Some issues that were encountered:
1. From outside the "sensitive" run, the command `pipe storage ls` works normally. But `pipe storage ls <Path_to_SensitiveStorage>` doesn't list the sensitive storage:
![image](https://user-images.githubusercontent.com/45459424/87606462-b395f500-c703-11ea-8ad3-d663ecb5e94c.png)
2. From inside the "sensitive" run, the command `pipe storage mount -b` doesn't allow to mount any storage (including sensitive):
![image](https://user-images.githubusercontent.com/45459424/87607118-471bf580-c705-11ea-88ed-b4546a9eb186.png)

**Comment by mzueva** [comment]
@NShaforostov Both issues shall be fixed in develop, could you please verify?

**Comment by NShaforostov** [comment]
@mzueva 
I've tried to verify.
Item 1 is fixed and works correctly now.

But I can't verify item 2 - now, **any** runs with "limit mounts" that contain sensitive storages are failed (I've tried `Ubuntu`, `Centos`, `RStudio`):
![image](https://user-images.githubusercontent.com/45459424/88163908-013ac200-cc1c-11ea-8446-77eeb8f83150.png)

**Comment by mzueva** [comment]
@NShaforostov This is fixed as well

**Comment by NShaforostov** [comment]
Verified, item 2 also works correctly.

**Comment by NShaforostov** [comment]
Docs were added via #2547 and can be found [here](https://github.com/epam/cloud-pipeline/blob/develop/docs/md/manual/08_Manage_Data_Storage/8.11._Sensitive_storages.md).
